### PR TITLE
Add ROM Header struct and ROM info dialog

### PR DESCRIPTION
**Pull request by WaluigiWare64**
**To-do list:**
- [x] Separate into one icon buffer and one index buffer.
- [x] Don't use seperate loops

image

### CODE REVIEW

File: src/frontend/Util_ROM.cpp [hunk 1]
```diff
@@ -461,6 +461,37 @@ int LoadROM(const char* file, int slot)
     }
 }
 
+u32* RomIcon(u8 data[512], u16 palette[16])
```
Comment (RSDuck) [inline src/frontend/Util_ROM.cpp:464]: if the data size is known to the caller and never too large, it's preferable for the caller to pass in a pointer where the data can be stored so the stack can be used for this.

File: src/NDSCart.cpp [hunk 2]
```diff
@@ -58,6 +58,9 @@ u32 CartID;
 bool CartIsHomebrew;
 bool CartIsDSi;
 
+struct NDSHeader Header;
```
Comment (RSDuck) [inline src/NDSCart.cpp:61]: writing struct before a struct type shouldn't be necessary in C++, unless it's different with packed one?

File: src/frontend/Util_ROM.cpp [hunk 3]
```diff
@@ -461,6 +461,37 @@ int LoadROM(const char* file, int slot)
...
+{
+    // Get the 4-bit palette indexes
+    u8 indexes[1024];
+    for (int i = 0; i < 512; i++)
```
Comment (RSDuck) [inline src/frontend/Util_ROM.cpp:468]: you don't have to use three loops for this, the entire operation can be done at once.

**Review by RSDuck — CHANGES_REQUESTED** [review]

File: src/frontend/Util_ROM.cpp (hunk 3 above)
Comment (WaluigiWare64) [reply src/frontend/Util_ROM.cpp:468]: Yeah i didn't write this, this is Hydr8gon's code that you showed me.

Comment (Arisotura) [reply src/frontend/Util_ROM.cpp:468]: Hydr8gon is a derp

File: src/frontend/Util_ROM.cpp (hunk 1 above)
Comment (Arisotura) [reply src/frontend/Util_ROM.cpp:464]: RomIcon? the capitalization there is inconsistent with the rest of the code

File: src/NDSCart.cpp (hunk 2 above)
Comment (Arisotura) [reply src/NDSCart.cpp:61]: shrug idk how all this struct shit works and why everybody does it differently

File: src/frontend/Util_ROM.cpp (hunk 1 above)
Comment (WaluigiWare64) [reply src/frontend/Util_ROM.cpp:464]: > 
> 
> RomIcon? the capitalization there is inconsistent with the rest of the code

Should I change it to ROMIcon?

File: src/frontend/Util_ROM.cpp [hunk 4]
```diff
@@ -461,6 +461,76 @@ int LoadROM(const char* file, int slot)
...
+        ROMIcon(data[SEQ_BMP(sequence[i])], palette[SEQ_PAL(sequence[i])], &animatedTexRef[32 * 32 * i]);
+
+        if (SEQ_FLIPH(sequence[i]))
+        {
+            for (int x = 0; x < 32; x++)
+            {
+                for (int y = 0; y < 32/2; y++) 
+                {
+                    int temp = frame[x * 32 + y];
+                    frame[x * 32 + y] = frame[x * 32 + (32 - 1 - y)];
```
Comment (RSDuck) [inline src/frontend/Util_ROM.cpp:511]: use std::swap for that

File: src/frontend/Util_ROM.cpp [hunk 5]
```diff
@@ -461,6 +461,76 @@ int LoadROM(const char* file, int slot)
...
+#define SEQ_DUR(i) ((i & 0b0000000011111111) >> 0)
+
+void AnimatedROMIcon(u8 (&data)[8][512], u16 (&palette)[8][16], u16 (&sequence)[64], u32 (&animatedTexRef)[32 * 32 * 64], std::vector<int> &animatedSequenceRef)
+{
+    for (int i = 0; i < 64; i++)
+    {
+        if (!sequence[i])
+            break;
+        u32* frame = &animatedTexRef[32 * 32 * i];
+        ROMIcon(data[SEQ_BMP(sequence[i])], palette[SEQ_PAL(sequence[i])], &animatedTexRef[32 * 32 * i]);
```
Comment (RSDuck) [inline src/frontend/Util_ROM.cpp:502]: replace &animatedTexRef[32 * 32 * i] by frame

**Pull request merged**